Update CRC32.c

benjaminkravets opened the pull request:
Add CheckCRC implementation

benjaminkravets committed:
Update CRC32.c

Add CheckCRC implementation

brian-amos-embedded commented:
Thanks for the PR Benjamin.  I'll merge it as soon as I have a chance to test it out!

benjaminkravets committed:
Update CRC32.c

I am adding an edit to this PR that uses the HAL provided CRC module (although I can't tell if the CRC module is initialized in the original project). I put a block-comment copy of my CRC initilization structure at the bottom. Not sure which of these approaches is better for the repository. Thank you for taking a look at my PR!

brian-amos-embedded commented:
Hi @benjaminkravets  - this change doesn't compile for me.  When I attempt to build Chapter_13 configuration *colorSelector* I get the following error:

````
Src/CRC32.o: In function `CheckCRC':
C:/projects/Hands-On-RTOS-with-Microcontrollers/Chapter_13/colorSelector/../Src/CRC32.c:21: undefined reference to `HAL_CRC_Calculate'
C:/projects/Hands-On-RTOS-with-Microcontrollers/Chapter_13/colorSelector/../Src/CRC32.c:44: undefined reference to `hcrc'
````
Based on those errors, there are two initial issues:
1. There are probably some missing files like `stm32f7xx_hal_crc.c` and `stm32f7xx_hal_crc.h` which need to be added to the *Hands-On-RTOS-with-Microcontrollers\Drivers\STM32F7xx_HAL_Driver\Src* and *Hands-On-RTOS-with-Microcontrollers\Drivers\STM32F7xx_HAL_Driver\Inc* directories, respectively.
2. `hcrc` is referenced as an extern, but no other files were in your commit that contain the definition of that variable.

I added the necessary files and the `#include` but was greeted with hundreds more errors due to some include chain issues.

What setup are you using to get this to compile?

benjaminkravets committed:
fix missing hcrc and missing packages

benjaminkravets commented:
Hey @brian-amos-embedded , thank you for taking a look! I added a commit that _should_ compile. I added the stm32f7xx_hal_crc.c and stm32f7xx_hal_crc.h you mentioned, stm32f7xx_hal_crc_ex.c, stm32f7xx_ll_crc.h, and stm32f7xx_hal_crc_ex.h. I also added 2 crc HAL initialization functions (HAL_CRC_MspInit and HAL_CRC_MspDeInit) to stm32f7xx_hal_msp.c, and finally enabled the CRC module in stm32f7xx_hal_conf and created an initialization function (MX_CRC_Init) similar to the ones you have in Nucleo_F767ZI_Init.c. I didn't put this in with those because I wanted local access to the hcrc pointer to pass to the CheckCRC function (just added it as an arg rather than referencing it externally), this could easily be refactored to go with the organization you have, just wasn't sure the best way to this.

As for what setup I am using, I am in STM32CubeIDE version 1.13.2 in Windows 11. When I was working through the book, I used CubeMX with the generated .ioc file to add the CRC package and the USB package (Pinout & Configuration > Categories > Computing > CRC).


As a footnote, one minor change I needed for my color selector script was uncapitalizing the PyCRC import, so it became 

`from pycrc.CRC32 import CRC32`

For some reason, when capitalized, Python3 couldn't find it. this may just be an issue with my environment, so I didn't commit it (on this front I am in Python 3.12.0).

Let me know if this works.
